Fix tile labels, add sunrise/sunset and report sensor failure

dacarson opened the pull request:
This fork has three fixes on it:
 #279 Report failing sensors. Shows as a general failure in app, also logs failure
 #247 Make use of ConfiguredName for compatibility values, to show values in the Home app where a native sensor isn't available
 #283 Report Sunset/Sunrise in forecast and current conditions when using Openweather API 2.5, and report Sunset/Sunrise for current conditions when using Openweather 3.0 API

Tested by running in my setup for a week, and 3.0 API was tested as documented here: https://github.com/jkreileder/homebridge-weather-plus/commit/e320cb91fd1499ea6773ce5d90d01edd63235696#r138496941

naofireblade commented:
Amazing! Thank you for your continuous improvement of the plugin. I will try to check and merge both new PRs on Monday.

naofireblade commented:
Sorry I had a busy week. Is this PR ready to merge or are u working on it again?

wimleers commented:
> #247 Make use of ConfiguredName for compatibility values, to show values in the Home app where a native sensor isn't available

Can you please move this out of this PR? This is a high-risk change, that only helps a subset of users. I have seen no naming issues.

dacarson commented:
I can remove the changes for #247  out the PR. Though I do struggle to see that it only affects a subset of users and that it is a high risk issue. Can you try enabling setting 'Compatibility' to Apple Home (plus Eve)?

When I did that, I saw that compatibility mode simply didn't work. As described [here](https://github.com/naofireblade/homebridge-weather-plus/issues/247#issuecomment-1375039404) when sensor names are updated, they are not updated in the app. There is a whole series of reports of "compatibility" mode not working.

With respect to high risk, I also beg to differ. Here is a [link](https://github.com/naofireblade/homebridge-weather-plus/pull/285/commits/f367473a173ed1d107a3922adfd74b3a5c888922) to the PR that has the change. The code that is impacted is only hit when you have Apple Home (plus Eve) compatibility mode set.

It seems valuable to have Apple Home (plus Eve) compatibility mode working.

wimleers commented:
Aha, interesting!

Upon a second reading of the code, I think you're right that the risk is limited.

(Note that I cannot easily switch to a different mode, because that'd break all of my existing automations that depend on the current services+characteristics 😅)

I'll get this up and running on my local instance to trial it 👍

dacarson commented:
You can add a separate weather station. I have one of each setup. Though the config / setup UI file needs to be fixed to make it work with multiple stations. I fixed the config file [here](https://github.com/naofireblade/homebridge-weather-plus/pull/285/commits/0fad356c1990459ed9c2c2a4f714bb51833ec761).
The problem was all subsequent weather stations adopted the settings of the first. The index numbering was always 0 rather than the current weather station. There was also a typo that broke the ability to save thresholds (thresholds are used in compatibility mode).

wimleers commented:
5165f22 crashes upon boot:
```
[2/23/2024, 9:28:09 PM] [homebridge-weather-plus] Loaded homebridge-weather-plus v3.3.3 child bridge successfully
[2/23/2024, 9:28:09 PM] Loaded 0 cached accessories from cachedAccessories.0EC69EE59806.
[2/23/2024, 9:28:09 PM] [homebridge-weather-plus] Adding station with weather service TempestAPI named 'Now'
[2/23/2024, 9:28:09 PM] [homebridge-weather-plus] Restoring last readings
[2/23/2024, 9:28:09 PM] Initializing platform accessory 'Now'...
[2/23/2024, 9:28:09 PM] [homebridge-weather-plus] server listening 0.0.0.0:50222
[2/23/2024, 9:28:09 PM] Homebridge v1.7.0 (HAP v0.11.1) (homebridge-weather-plus) is running on port 32685.

/Users/admin/Desktop/homebridge-weather-plus/apis/weatherflow.js:588
                                callback(parseError, weather);
    ^
TypeError: Cannot read properties of undefined (reading 'time')
    at /Users/admin/Desktop/homebridge-weather-plus/apis/weatherflow.js:216:74
    at Request._callback (/Users/admin/Desktop/homebridge-weather-plus/apis/weatherflow.js:588:5)
    at Request.self.callback (/Users/admin/Desktop/homebridge-weather-plus/node_modules/request/request.js:185:22)
    at Request.emit (node:events:514:28)
    at Request.<anonymous> (/Users/admin/Desktop/homebridge-weather-plus/node_modules/request/request.js:1154:10)
    at Request.emit (node:events:514:28)
    at IncomingMessage.<anonymous> (/Users/admin/Desktop/homebridge-weather-plus/node_modules/request/request.js:1076:12)
    at Object.onceWrapper (node:events:628:28)
    at IncomingMessage.emit (node:events:526:35)
    at endReadableNT (node:internal/streams/readable:1589:12)
[2/23/2024, 9:28:10 PM] [homebridge-weather-plus] Child bridge process ended
```

wimleers commented:
Note that this was happening with ZERO config changes. 

Ah, after configuring `stationId` and `key` (confusingly named for the Weatherflow "Personal Use Token" 😅, `personalUseToken` would have been far clearer), that crash no longer occurs. Great. But  that does mean that everybody who upgrades will run into this plugin crashing. It'd be much nicer if it didn't fail hard, but instead did a `this.log.warn()` and simply would not make that functionality work. Especially if you don't care about the forecasts (I didn't even enable them!).

dacarson commented:
Agreed. Let me fix that.

wimleers commented:
I'm seeing:
- Eve app:
![image](https://github.com/naofireblade/homebridge-weather-plus/assets/14240/0b70c3c1-bbca-4027-a49d-bfc554c15978)
- Home+ app: 
![image](https://github.com/naofireblade/homebridge-weather-plus/assets/14240/21dda80e-2492-4132-8f7e-e38e43ff1278)
- Controller app: 
![image](https://github.com/naofireblade/homebridge-weather-plus/assets/14240/71b1f955-71e0-47ec-b262-086c809714be)
- (official) Home app: nothing, because I'm not using the compatibility mode

☝️ Where is that "code 1" coming from? Can we customize that? (Clearly Eve surfaces this best! 👏)

---

And getting this in the logs:
```
 [ 2 / 2 3 / 2 0 2 4 ,   9 : 3 3 : 3 7   P M ]   [ h o m e b r i d g e - w e a t h e r - p l u s ]   S e n s o r   f a i l :   L i g h t n i n g   f a i l e d   L i g h t n i n g   n o i s e   L i g h t n i n g   d i s t u r b e r   P r e s s u r e   f a i l e d   T e m p e r a t u r e   f a i l e d   R e l a t i v e   H u m i d i t y   f a i l e d   W i n d   f a i l e d   P r e c i p i t a t i o n   f a i l e d   L i g h t / U V   f a i l e d   P o w e r   b o o s t e r   s h o r e   p o w e r   
 [ 2 / 2 3 / 2 0 2 4 ,   9 : 3 4 : 3 7   P M ]   [ h o m e b r i d g e - w e a t h e r - p l u s ]   S e n s o r   f a i l :   L i g h t n i n g   f a i l e d   L i g h t n i n g   n o i s e   L i g h t n i n g   d i s t u r b e r   P r e s s u r e   f a i l e d   T e m p e r a t u r e   f a i l e d   R e l a t i v e   H u m i d i t y   f a i l e d   W i n d   f a i l e d   P r e c i p i t a t i o n   f a i l e d   L i g h t / U V   f a i l e d   P o w e r   b o o s t e r   s h o r e   p o w e r  
```
☝️ I think formatting could be better. Not multiple lines. Just some punctuation would do wonders.

dacarson commented:
Unfortunately, there is only 'code 1' status fault recognized by HomeKit:
https://developer.apple.com/documentation/homekit/hmcharacteristicvaluestatusfault

Eve might use other numbers to give a more descriptive result, but I don't see anyone documenting those anywhere.

dacarson commented:
> https://github.com/naofireblade/homebridge-weather-plus/commit/5165f22a55baf24dc1e7cc6c0b3b254bee0f00e1 crashes upon boot:

Let me put in some code to make sure that it is resilient to missing or bad apiKey and/or stationId

wimleers commented:
Thanks for everything you're doing — you're truly pushing this functionality to the next level! 🤩

I just enabled forecasts. For today + tomorrow. This is what those look like:
![today](https://github.com/naofireblade/homebridge-weather-plus/assets/14240/eeea493f-d8c6-4a13-a341-084f5885421d)
![tomorrow](https://github.com/naofireblade/homebridge-weather-plus/assets/14240/21053412-7d1f-4fe6-b206-23c327161f57)

Observations:
- `Observation time` is listed for both but seems meaningless? Or is that the time it was based upon? Or maybe it's only fetched once per hour?
- Why `In 1 day` and not just `Tomorrow`? 🤨
- Why in the "Today" view does it say `Rain probability: 20%`, `Condition: Rain possible` (these make sense) and then `Rain: Yes`? I guess that's just because any non-zero chance would cause it to say `Yes`?

dacarson commented:
> * `Observation time` is listed for both but seems meaningless?

The way the code is structured, I need to return the same fields for each forecast day. So I can't list it for one day and not another.

> Or is that the time it was based upon? 

It is the time field provided in the [BetterForecastCurrentConditions](https://weatherflow.github.io/Tempest/api/swagger/#!/forecast/getBetterForecast). I believe it is the time of the forecast. This is the same as how OpenWeatherMap populates that field.  I think it is useful, it represent a time of when the forecast was made.

> Or maybe it's only fetched once per hour?

I do only fetch it once an hour as (a) it is rate limited API call and (b) I can't see it changing more frequently than that. This is hard coded because the refresh interval is set to one minute to refresh the UDP data, and I don't want to call the forecast API each minute.  

> * Why `In 1 day` and not just `Tomorrow`? 🤨

That decision was made before my time. Maybe it is for easier localization to other languages?

> * Why in the "Today" view does it say `Rain probability: 20%`, `Condition: Rain possible` (these make sense) and then `Rain: Yes`? I guess that's just because any non-zero chance would cause it to say `Yes`?

Correct.
I followed the behavior in OpenWeatherMap, which is if the weather condition is (possible) Rain, Drizzle or Storm, then it sets the `Rain` to `true`. Similarly if the weather condition is (possible) Snow or Hail, then it sets the `Snow` to `true`.
One difference is, I include Hail (Sleet) into the Rain checks and not the Snow checks. IMO that is better.

BTW The icon next to `Condition` in the Eve app does change depending on the forecast.

wimleers commented:
> Unfortunately, there is only 'code 1' status fault recognized by HomeKit: https://developer.apple.com/documentation/homekit/hmcharacteristicvaluestatusfault
> 
> Edit: NoFault == 0 and generalFault == 1
> 
> Eve might use other numbers to give a more descriptive result, but I don't see anyone documenting those anywhere.

If https://github.com/thoukydides/homebridge-homeconnect/issues/6#issuecomment-570773152 is accurate, you should be able to generate different error codes! 😊

wimleers commented:
> ```
> [2/23/2024, 9:41:37 PM] [homebridge-weather-plus] Sensor fail: Lightning failed Lightning noise Lightning disturber Pressure failed Temperature failed Relative Humidity failed Wind failed Precipitation failed Light/UV failed Power booster shore power 
> ```
> 
> ☝️ I think formatting could be better. Not multiple lines. Just some punctuation would do wonders.

Also, this is now appearing _once per minute_. This means >90% of my Homebridge log is this exact same line. Can we reduce the reporting so it only reports once per hour? (You can keep a trivial counter and do something like `let shouldWarnAboutFailures = observationsSinceLastFailure % 60 === 0;` to achieve that.) The message could then be updated to say `Will report again in one hour or after a restart of this Homebridge plugin.`

wimleers commented:
> The way the code is structured, I need to return the same fields for each forecast day. So I can't list it for one day and not another.

👍 

> It is the time field provided in the [BetterForecastCurrentConditions](https://weatherflow.github.io/Tempest/api/swagger/#!/forecast/getBetterForecast). I believe it is the time of the forecast. This is the same as how OpenWeatherMap populates that field. I think it is useful, it represent a time of when the forecast was made.

I see! I wish it was called `Forecast time` or `Forecast last updated time` or something like that then. But … per the previous point, that's not possible! So, yeah, it's kind of "observation time of this forecast", which is close enough 😄👍

> I do only fetch it once an hour as (a) it is rate limited API call and (b) I can't see it changing more frequently than that. This is hard coded because the refresh interval is set to one minute to refresh the UDP data, and I don't want to call the forecast API each minute.

Totally reasonable! 👍

> That decision was made before my time. Maybe it is for easier localization to other languages?

I see. 👍

> BTW The icon next to `Condition` in the Eve app does change depending on the forecast.

🤩 The Eve app just keeps surprising — I'm impressed/surprised they have icons for _so incredibly many_ non-official (i.e. not defined by the Apple HomeKit spec) characteristics! 😮

---

Thank you as always for your kind responses, your tireless work on this plugin. It is _deeply_ appreciated. If you ever make it to Belgium, let me know. I'll get you the best beers 😄 (alcoholic or non-alcoholic)!

dacarson commented:
> > Unfortunately, there is only 'code 1' status fault recognized by HomeKit: https://developer.apple.com/documentation/homekit/hmcharacteristicvaluestatusfault
> > Edit: NoFault == 0 and generalFault == 1
> > Eve might use other numbers to give a more descriptive result, but I don't see anyone documenting those anywhere.
> 
> If [thoukydides/homebridge-homeconnect#6 (comment)](https://github.com/thoukydides/homebridge-homeconnect/issues/6#issuecomment-570773152) is accurate, you should be able to generate different error codes! 😊

You are correct. What I could do is just return the sensor_status number directly if there is an error. You would still need to know that that is what the number is, but it might be better.

dacarson commented:
Summary of changes needed:
1. If possible, calculate calculated fields even if there is a sensor error
2. If there is a sensor error, return the sensor_status number directly as the StatusFault
3. Don't continuously log the sensor failure, log it every hour.
4. Make the code robust so that if apiKey and/or stationId values are missing or bad, don't crash.

wimleers commented:
With those additions/changes, the Tempest Weatherflow support would be more robust than ever before! 😄 I guess that's the silver lining in all three of us being affected by three different failure modes? 🙈 🤞

dacarson commented:
> 1. If possible, calculate calculated fields even if there is a sensor error

I checked the source code for the [weather formulas](https://github.com/oyve/weather-formulas) I am using and it will produce NAN/failures if the numbers are not within range. 
I modified the code to ensure values are within range (20c96ea)

dacarson commented:
> 2. If there is a sensor error, return the sensor_status number directly as the StatusFault

Tried this and I get an error message from HomeBridge. 
```
[2/23/2024, 4:46:26 PM] [WeatherPlus] Sensor fail: Lightning disturber 
[2/23/2024, 4:46:48 PM] [homebridge-weather-plus] This plugin generated a warning from the characteristic 'Status Fault': characteristic was supplied illegal value: number 4 exceeded maximum of 1. See https://homebridge.io/w/JtMGR for more info.
```
With the exception of that [one mention](https://github.com/thoukydides/homebridge-homeconnect/issues/6#issuecomment-570773152) in comments, I could not find any Homebridge extension that has code that returns a value other than 0 or 1.
I can override it and force the number through, but I am reluctant to do this as I don’t know the behavior in all apps

dacarson commented:
Also, thank you very much for the testing and the feedback @wimleers and @jmissig. It does make the support better.

dacarson commented:
> 3\. Don't continuously log the sensor failure, log it every hour.

Changed the logging to log initial failure and then every hour so that it doesn't spam the logs with failure entries (https://github.com/naofireblade/homebridge-weather-plus/pull/285/commits/b99b173f7ce7e9abe9ca7b1d41ad9e329dd59086). Logging is reset when the sensor status is set back to zero. However, as there can be AIR and SKY units used, and they both separately report sensor status, I need to keep track of which unit is failing so I can reset logging for the correct unit. Elsewhere in the code I overload the Sky unit with the Tempest unit as it is unlikely to have both, so I do the same in the error reporting. 

In this change, I also removed the variables that were tracking Temperature and Humidity sensor failures as those variables are no longer needed now that I am bounds checking the values.

dacarson commented:
> 4. Make the code robust so that if apiKey and/or stationId values are missing or bad, don't crash.

There are two changes needed here (https://github.com/naofireblade/homebridge-weather-plus/pull/285/commits/db03a9ecfaf8a8b57d13a77e6e8538bd80431a80). 
First change is to not attempt to fetch the forecast if the apiKey and/or stationId is not set. I use the lastForecastUpdate variable to indicate if we have valid parameters at all. If they are not set, don't define the variable, and as such never attempt to fetch a forecast. This is a valid configuration so I don't log this state. 
The second change is to handle invalid or incorrect apiKey and/or stationId values. I now check the URL response and fail if the response code is anything other than 200 (OK). This error will get logged. As the forecast is only fetched every hour, this error will only appear once an hour and won't spam the logs.

wimleers commented:
https://github.com/naofireblade/homebridge-weather-plus/commit/20c96ea5b808b412ef94e488e5bb664ebeec645d specified `0` as the minimum temperature. That’s true in Kelvin (“absolute zero”), but not in Celsius, where negative degrees are freezing temperatures.

wimleers commented:
It seems that ever since installing this branch (since https://github.com/naofireblade/homebridge-weather-plus/pull/285#issuecomment-1962003233), my temperature history, light level and air pressure history are no longer updating? 🤔 

![image](https://github.com/naofireblade/homebridge-weather-plus/assets/14240/2f6f18b6-76e5-43bc-99f8-b5ace7418fa6)
![image](https://github.com/naofireblade/homebridge-weather-plus/assets/14240/73022c71-89dd-45db-83e1-f2fc10479521)

wimleers commented:
Update: works fine now, with no missing history 🤔 Seems like a glitch in the Eve app? I did restart it before reporting 🤷

dacarson commented:
> [20c96ea](https://github.com/naofireblade/homebridge-weather-plus/commit/20c96ea5b808b412ef94e488e5bb664ebeec645d) specified `0` as the minimum temperature. That’s true in Kelvin (“absolute zero”), but not in Celsius, where negative degrees are freezing temperatures.

I realized this too, yesterday. I changed it in my internal code to -100 minimum. It'll be fixed in the next PR.

dacarson commented:
> 6. Make device_status more robust, and validate serial_number against observation before applying info from it.

I actually took care of this in (https://github.com/naofireblade/homebridge-weather-plus/commit/b99b173f7ce7e9abe9ca7b1d41ad9e329dd59086) where I separated the error handling of AIR vs SKY. 

While testing with separate SKY/AIR units, I found that the battery calculations are wrong:
```
[2/26/2024, 11:00:52 AM] [homebridge-weather-plus] This plugin generated a warning from the characteristic 'Battery Level': characteristic was supplied illegal value: number -17 exceeded minimum of 0. See https://homebridge.io/w/JtMGR for more info.
[2/26/2024, 11:01:52 AM] [homebridge-weather-plus] This plugin generated a warning from the characteristic 'Battery Level': characteristic was supplied illegal value: number -16 exceeded minimum of 0. See https://homebridge.io/w/JtMGR for more info.
[2/26/2024, 11:02:52 AM] [homebridge-weather-plus] This plugin generated a warning from the characteristic 'Battery Level': characteristic was supplied illegal value: number -16 exceeded minimum of 0. See https://homebridge.io/w/JtMGR for more info.
```
So I need to add one more change to this fork. 
7. Fix SKY/AIR battery calculations so that they must be within 0 - 100 % **and** as there can only be one battery level, report the lowest value. As the user cares about low battery.

dacarson commented:
> 7. Fix SKY/AIR battery calculations so that they must be within 0 - 100 % **and** as there can only be one battery level, report the lowest value. As the user cares about low battery.

Fix (https://github.com/naofireblade/homebridge-weather-plus/pull/285/commits/6f2e152fbdfc656c2dd1d35d4e0a2dde9df0fc69). Remove partial duplicate handling of battery level from device_status message, and put all battery handling in the observation handling code. This was already the case for Tempest, just not the case for AIR/SKY sensors. As there can only be one battery level reported, but we may have AIR & SKY units, pick the lowest value and report it.
Also switch the battery calculations for AIR/SKY to be the same as Tempest as the ranges and values seem close enough in observation.

dacarson commented:
@wimleers all feedback has been applied 😀 
Please try out this version and hit me up with questions and/or feedback.

wimleers commented:
Deployed and testing 😊 Will report back in ~24 hours!

wimleers commented:
Good news from Weatherflow support:
> We wouldn't require you to perform a manual installation for a beta firmware. Once we fixed the issue, we should update the firmware for your Tempest device and the sensor flags should disappear. I apologize for the inconvenience!

wimleers commented:
> Deployed and testing 😊 Will report back in ~24 hours!

All is well!

Only one odd thing: the `Observation time` is seemingly not updating — it's still saying `00:00:15` while it's 19:25. That suggests the forecast is not being updated?

dacarson commented:
> Only one odd thing: the `Observation time` is seemingly not updating — it's still saying `00:00:15` while it's 19:25. That suggests the forecast is not being updated?

If everything else is updating, the I wonder if it is a glitch in the Eve app. There is no reason for it to not update.

wimleers commented:
> > Only one odd thing: the `Observation time` is seemingly not updating — it's still saying `00:00:15` while it's 19:25. That suggests the forecast is not being updated?
> 
> If everything else is updating, the I wonder if it is a glitch in the Eve app. There is no reason for it to not update.

Worse, it doesn’t update at all anymore! I bet the recent refactoring broke it. Or the config key was renamed and I missed the log entry. Restarting.

wimleers commented:
Nothing:
```
545577181065706 high at 231.106451200303
[2/28/2024, 4:13:47 PM] [homebridge-weather-plus] Restarting child bridge...
[2/28/2024, 4:13:47 PM] Got SIGTERM, shutting down child bridge process...
[2/28/2024, 4:13:52 PM] [homebridge-weather-plus] Child bridge process ended
[2/28/2024, 4:13:52 PM] [homebridge-weather-plus] Process Ended. Code: 143, Signal: null
[2/28/2024, 4:13:59 PM] [homebridge-weather-plus] Restarting Process...
[2/28/2024, 4:13:59 PM] [homebridge-weather-plus] Launched child bridge with PID 14905
[2/28/2024, 4:14:00 PM] Registering platform 'homebridge-weather-plus.WeatherPlus'
[2/28/2024, 4:14:00 PM] [homebridge-weather-plus] Loaded homebridge-weather-plus v3.3.3 child bridge successfully
[2/28/2024, 4:14:00 PM] Loaded 0 cached accessories from cachedAccessories.0EC69EE59806.
[2/28/2024, 4:14:00 PM] [homebridge-weather-plus] Adding station with weather service TempestAPI named 'Now'
[2/28/2024, 4:14:00 PM] [homebridge-weather-plus] Restoring last readings
[2/28/2024, 4:14:00 PM] Initializing platform accessory 'Now'...
[2/28/2024, 4:14:00 PM] Initializing platform accessory 'Today'...
[2/28/2024, 4:14:00 PM] Initializing platform accessory 'In 1 Day'...
[2/28/2024, 4:14:00 PM] [homebridge-weather-plus] server listening 0.0.0.0:50222
[2/28/2024, 4:14:00 PM] Homebridge v1.7.0 (HAP v0.11.1) (homebridge-weather-plus) is running on port 32685.
[2/28/2024, 4:14:01 PM] [homebridge-weather-plus] This plugin generated a warning from the characteristic 'Air Pressure': characteristic was supplied illegal value: number 0 exceeded minimum of 700. See https://homebridge.io/w/JtMGR for more info.
```

dacarson commented:
If you continually see this line appearing in your logs:

> `[2/28/2024, 4:14:01 PM] [homebridge-weather-plus] This plugin generated a warning from the characteristic 'Air Pressure': characteristic was supplied illegal value: number 0 exceeded minimum of 700. See https://homebridge.io/w/JtMGR for more info.`

It is because the plugin has not received any data from the weather station. (The error occurs as I initialize the Air Pressure value to zero, and is only changed when it receives some data from the weather station). I do expect it could appear once on startup because no data has been received yet, but the plugin has requested initial values.

To verify that the plugin is receiving data, you can change line 149:
`this.log.debug(`Server got: ${message.type}`);`
to this
`this.log.debug(`Server got: ${msg.toString()}`);`

FWIW, with separate AIR/SKY sensors, everything is working fine for me.

wimleers commented:
I figured out why forecasts aren't updating:
```

/Users/admin/Desktop/homebridge-weather-plus/apis/weatherflow.js:637
					callback(parseError, weather);
     ^
TypeError: Cannot read properties of undefined (reading 'time')
    at /Users/admin/Desktop/homebridge-weather-plus/apis/weatherflow.js:222:74
    at Request._callback (/Users/admin/Desktop/homebridge-weather-plus/apis/weatherflow.js:637:6)
    at Request.self.callback (/Users/admin/Desktop/homebridge-weather-plus/node_modules/request/request.js:185:22)
    at Request.emit (node:events:514:28)
    at Request.<anonymous> (/Users/admin/Desktop/homebridge-weather-plus/node_modules/request/request.js:1154:10)
    at Request.emit (node:events:514:28)
    at IncomingMessage.<anonymous> (/Users/admin/Desktop/homebridge-weather-plus/node_modules/request/request.js:1076:12)
    at Object.onceWrapper (node:events:628:28)
    at IncomingMessage.emit (node:events:526:35)
    at endReadableNT (node:internal/streams/readable:1589:12)
[2/28/2024, 10:00:02 PM] [homebridge-weather-plus] Child bridge process ended
[2/28/2024, 10:00:02 PM] [homebridge-weather-plus] Process Ended. Code: 1, Signal: null
[2/28/2024, 10:00:09 PM] [homebridge-weather-plus] Restarting Process...
[2/28/2024, 10:00:10 PM] [homebridge-weather-plus] Launched child bridge with PID 75453
```

wimleers commented:
> To verify that the plugin is receiving data, you can change line 149:

Data _is_ being received. The Eve history view for `Air Pressure` works fine. 👍 

But I'm seeing occasional drops it seems: 
<img width="900" alt="image" src="https://github.com/naofireblade/homebridge-weather-plus/assets/14240/2c69ade7-054e-440c-9466-946e09562d58">

dacarson commented:
> I figured out why forecasts aren't updating:

Can you share your JSON config? 

This error:
`TypeError: Cannot read properties of undefined (reading 'time')`
happens when it is trying to load the forecast, but the returned JSON is not a forecast. I thought I had fixed that by checking the http status code **before** trying to read the forecast (https://github.com/naofireblade/homebridge-weather-plus/commit/db03a9ecfaf8a8b57d13a77e6e8538bd80431a80), as well as not requesting the forecast if the apiKey and/or stationID was not provided.

Obviously, it isn't robust enough. To understand why it is failing your JSON would help.

wimleers commented:
> I figured out why forecasts aren't updating:
>
> …

I suspect https://github.com/dacarson/homebridge-weather-plus/commit/db03a9ecfaf8a8b57d13a77e6e8538bd80431a80 caused this regression? (I find `node.js` backtraces notoriously difficult to read, due to its infamous "callback hell".)

Related: it might be worth doing a `this.debug.info(…)` whenever the forecast is updated — that'd result in 24 log entries per day, that seems reasonable? I'd suggest `Successfully fetched a new forecast …` for a `200` response and a `this.log.error(…)` with `Failed to fetch new forecast. Got response with status code 500 containing: <raw response body>` for any 4xx or 5xx response? A non-200 response should also cause a Status Fault for a forecast accessory?

dacarson commented:
> But I'm seeing occasional drops it seems:

My initial thought is that the value is reset, but I never actually reset a value. If a new value has not be received, then I just send the previous value. 
The only time the value is 'reset' is when the plugin is restarted and it has not received any data from the weather station yet. Thinking about this further, what I could do in this case is return an error when the data is [requested](https://github.com/dacarson/homebridge-weather-plus/blob/679d48befed1d7b1af32093e4de5dc863db8ae54/index.js#L248). When a weather station returns an error, it doesn't update the values **and** it doesn't update the history.

wimleers commented:
> The only time the value is 'reset' is when the plugin is restarted and it has not received any data from the weather station yet.

It's totally possible that this edge case occurs whenever A) there is an intentional Homebridge (child bridge) restart, B) it just happens to be within that 59 second time window where no data has been received yet, C) a crash like the one we're discussing in the surrounding comments is causing such restarts! 😄 🤓

dacarson commented:
I received the JSON config and I put it in my config. I am not seeing the error yet. Though to avoid the crash, and help debugging, I made this change to forecast fetching (https://github.com/naofireblade/homebridge-weather-plus/pull/285/commits/37409db2a05b04e2b283abad6b35533e36d64e5d). If it hits this issue again, it will log the response so that I can debug it 😊

wimleers commented:
Thursday, just before midnight, looking at the forecast for tomorrow:
![IMG_7855](https://github.com/naofireblade/homebridge-weather-plus/assets/14240/de61f68d-1877-42a2-b2bd-c95e8895dc83)

Friday, well past midnight:
![IMG_7858](https://github.com/naofireblade/homebridge-weather-plus/assets/14240/f3e5131f-f0ab-4f84-a97f-dd3273465c56)

No log entries at all. The crash from the other day hasn’t occurred since. I suspect it was an ephemeral server issue over at Weatherflow.

But this problem seems to lie in the plugin.

dacarson commented:
This problem is because I only update once an hour. The previous forecast 23:59 and when you checked again at 00:30, an hour has not passed so the forecast has not updated. If you checked at 00:59 or 01:00 I believe it'll have a new forecast.

wimleers commented:
> This problem is because I only update once an hour. The previous forecast 23:59 and when you checked again at 00:30, an hour has not passed so the forecast has not updated. If you checked at 00:59 or 01:00 I believe it'll have a new forecast.

Good theory! 😄  But unfortunately the Eve app (restarted this morning just to make sure) confirms what I saw earlier this week: that it just doesn't update _at all_:
![image](https://github.com/naofireblade/homebridge-weather-plus/assets/14240/02695d18-29f1-46c5-90d2-004daa9ab405)

The only log entries from `homebridge-weather-plus` are the hourly ones that report my sensor failures.

Now that I've gathered that data: updating to https://github.com/naofireblade/homebridge-weather-plus/commit/37409db2a05b04e2b283abad6b35533e36d64e5d

wimleers commented:
It's now Saturday. I just opened the Eve app on my iPad for the first time in days. I see the exact same information as in the screenshot I posted in the Friday AM. The "Tomorrow" prediction is now the prediction fetched on Thursday about Friday. So it's "forecasting" yesterday 😅

dacarson commented:
And no errors are appearing in the log?

jmissig commented:
What's the easiest way for me to try this branch at home? I tried a `git clone` of @dacarson's fork and then an `npm -g install` from within that directory, but Homebridge complains about a bunch of missing stuff until I `npm -g install homebridge-weather-plus` to go back to main. I'm guessing I'm missing a step somewhere.

dacarson commented:
I actually `npm -g install homebridge-weather-plus` then copy the cloned files over the top of the installed files.

dacarson commented:
Oddly without any intervention, my tempest unit suddenly woke up again. ![image](https://github.com/naofireblade/homebridge-weather-plus/assets/44933987/8d94c55e-51bb-47e7-84f8-8e5b455c537d)

wimleers commented:
> And no errors are appearing in the log?

None.

dacarson commented:
> > And no errors are appearing in the log?
> 
> None.

I have been traveling, so I haven't been able to work on this till today.

Forecast stopped updating for me too. I think I know what the issue is. It is this line:
`if (this.lastForecastUpdate && moment().hour() != this.lastForecastUpdate) {`
What I wanted to happen was, if `lastForecastUpdate` was not defined, then don't even attempt to load the forecast. The variable `lastForecastUpdate` was only defined if an API key and Station ID was provided.

What I think is happening, when the hour clicked over midnight, the current hour is `0`, the `if` statement returns `false` and the forecast is never updated again.

The test I am using for an undefined variable is incorrect. It looks like what I should be doing is:
`if ((typeof this.lastForecastUpdate !== 'undefined')  && moment().hour() != this.lastForecastUpdate) {`
I am testing this change out.

dacarson commented:
> * Why in the "Today" view does it say `Rain probability: 20%`, `Condition: Rain possible` (these make sense) and then `Rain: Yes`? I guess that's just because any non-zero chance would cause it to say `Yes`?

I did find an issue with this. As can be seen in my forecast below ,  `Rain Probability: 0%`, and `Condition: Partly Cloudy`, but it still showed `Rain: Yes`. I was basing the `Rain` Yes/No on the precipitation icon, which was incorrect as there will always be a precipitation icon even when rain/snow is not in the forecast. A better way for `Rain` Yes/No was to copy exactly what OpenWeatherMap is doing. They normalize the weather condition into one of the conditions that Eve understand, then they check if the condition is rain or snow. 
![IMG_5046](https://github.com/naofireblade/homebridge-weather-plus/assets/44933987/aa8079e3-0acf-424f-b3cb-9eaeaef8489c)

dacarson commented:
@wimleers & @jmissig Could you test this again. I fixed the issue with Forecast not updating as well as a fix around Rain in forecast.

wimleers commented:
Updated! Will be keeping an eye on this.

P.S.: isn't it funny that the forecast seemed to be working fine for you, but then only thanks to _not working on this_ and hence letting it run without interruption for multiple days, you were able to reproduce it? 😄

dacarson commented:
> P.S.: isn't it funny that the forecast seemed to be working fine for you, but then only thanks to _not working on this_ and hence letting it run without interruption for multiple days, you were able to reproduce it? 😄

It was 'working for me' because there was another issue I had in my Homebridge that caused it to restart once a day. Restarting, got me past the zero hour issue, and I didn't see it initially. I fixed my restart issue, and that is when I experienced the same issue you were.

dacarson commented:
My internet went down for a short period of time, got errors from other plugins but HomeBridge ended up restarting because of a change I made here https://github.com/naofireblade/homebridge-weather-plus/pull/285/commits/67c11a604a17f9ad9b4a9ca0e554cd384aa3e251. 
Patch coming

```
[07/03/2024, 18:04:58] [B-Hyve] Error getting API key getaddrinfo EAI_AGAIN api.orbitbhyve.com
[07/03/2024, 18:04:58] [B-Hyve] Error retrieving API key 
TypeError: Cannot read properties of undefined (reading 'status')
    at OrbitAPI.getToken (/var/lib/homebridge/node_modules/homebridge-orbit-irrigation/orbitapi.js:47:17)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at OrbitPlatform.getDevices (/var/lib/homebridge/node_modules/homebridge-orbit-irrigation/orbitplatform.js:89:24)
[07/03/2024, 18:04:58] [B-Hyve] Failed to get devices. Retry attempt 1 of 3 in 60 seconds...
[07/03/2024, 18:04:58] [B-Hyve] TypeError: Cannot read properties of undefined (reading 'user_name')
    at OrbitPlatform.getDevices (/var/lib/homebridge/node_modules/homebridge-orbit-irrigation/orbitplatform.js:90:53)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
Error: getaddrinfo EAI_AGAIN api-production.august.com
    at GetAddrInfoReqWrap.onlookupall [as oncomplete] (node:dns:118:26) {
  errno: -3001,
  code: 'EAI_AGAIN',
  syscall: 'getaddrinfo',
  hostname: 'api-production.august.com'
}
[07/03/2024, 18:05:09] [Homebridge UI] [homebridge] Failed to check registry.npmjs.org for updates: "timeout of 10000ms exceeded" - see https://homebridge.io/w/JJSz6 for help.
[07/03/2024, 18:05:09] [Homebridge UI] [homebridge-orbit-irrigation] Failed to check registry.npmjs.org for updates: "timeout of 10000ms exceeded" - see https://homebridge.io/w/JJSz6 for help.
[07/03/2024, 18:05:09] [Homebridge UI] [homebridge-weather-plus] Failed to check registry.npmjs.org for updates: "timeout of 10000ms exceeded" - see https://homebridge.io/w/JJSz6 for help.
[07/03/2024, 18:05:09] [Homebridge UI] [homebridge-orbit-irrigation] Failed to check registry.npmjs.org for updates: "timeout of 10000ms exceeded" - see https://homebridge.io/w/JJSz6 for help.
[07/03/2024, 18:05:09] [Homebridge UI] [homebridge-weather-plus] Failed to check registry.npmjs.org for updates: "timeout of 10000ms exceeded" - see https://homebridge.io/w/JJSz6 for help.
[07/03/2024, 18:05:18] [WeatherPlus] Error retrieving weather Forecast
[07/03/2024, 18:05:18] [WeatherPlus] undefined
[07/03/2024, 18:05:18] [WeatherPlus] Error getting locationGeo from London, GB
[07/03/2024, 18:05:18] [WeatherPlus] Error: getaddrinfo EAI_AGAIN api.openweathermap.org
    at GetAddrInfoReqWrap.onlookupall [as oncomplete] (node:dns:118:26) {
  errno: -3001,
  code: 'EAI_AGAIN',
  syscall: 'getaddrinfo',
  hostname: 'api.openweathermap.org'
}
[07/03/2024, 18:05:19] [Homebridge UI] [homebridge-config-ui-x] Failed to check registry.npmjs.org for updates: "timeout of 10000ms exceeded" - see https://homebridge.io/w/JJSz6 for help.
[07/03/2024, 18:05:19] [Homebridge UI] [homebridge-config-ui-x] Failed to check registry.npmjs.org for updates: "timeout of 10000ms exceeded" - see https://homebridge.io/w/JJSz6 for help.
[07/03/2024, 18:05:38] [WeatherPlus] Weather Underground Request failed
[07/03/2024, 18:05:38] TypeError: Cannot read properties of undefined (reading 'statusCode')
    at WundergroundAPI.<anonymous> (/var/lib/homebridge/node_modules/homebridge-weather-plus/apis/weatherunderground.js:73:55)
    at self.callback (/var/lib/homebridge/node_modules/homebridge-weather-plus/node_modules/request/request.js:185:22)
    at Request.emit (node:events:518:28)
    at Request.onRequestError (/var/lib/homebridge/node_modules/homebridge-weather-plus/node_modules/request/request.js:877:8)
    at ClientRequest.emit (node:events:518:28)
    at TLSSocket.socketErrorListener (node:_http_client:495:9)
    at TLSSocket.emit (node:events:518:28)
    at emitErrorNT (node:internal/streams/destroy:169:8)
    at emitErrorCloseNT (node:internal/streams/destroy:128:3)
    at processTicksAndRejections (node:internal/process/task_queues:82:21)
[07/03/2024, 18:05:38] Got SIGTERM, shutting down Homebridge...
```

wimleers commented:
It's been working well now! 😄

Just spotted one more (and probably/hopefully last) bug though: the _name of the day_  for a forecast appears to be computed using UTC or the non-DST-adjusted time.

- Friday, 23:20: ![IMG_7952](https://github.com/naofireblade/homebridge-weather-plus/assets/14240/fe376bf2-0d0b-45a9-aa3c-05f4d5fb84ad)
- Saturday, 00:12: ![IMG_7954](https://github.com/naofireblade/homebridge-weather-plus/assets/14240/f9460891-c0f0-4851-9c4c-fe0033735348)

dacarson commented:
> Just spotted one more (and probably/hopefully last) bug though: the _name of the day_ for a forecast appears to be computed using UTC or the non-DST-adjusted time.

DST is a tricky one. Sunrise and Sunset are computed using the same algorithm. Are they off by an hour too?
```
forecast.ForecastDay =  moment.unix(values[i].day_start_local).tz(timezone).format('dddd');

forecast.SunriseTime = moment.unix(values[i].sunrise).tz(timezone).format('HH:mm:ss');
forecast.SunsetTime = moment.unix(values[i].sunset).tz(timezone).format('HH:mm:ss');
```

The [time library](https://momentjs.com/timezone/docs/#/using-timezones/parsing-ambiguous-inputs/) that is being used should automatically take DST into account. 

The `Observation` time  in the screenshots are both before midnight: 22:59:36, and 23:59:36 so it thinks it is still on the same day. This will happen regardless of DST. On the next update, at 00:59:36, it should click over to the next day.

I could see if I can change the forecast fetching to the beginning of every hour rather than the end of every hour, that way as soon as it ticks over midnight, the forecast for Today, Tomorrow, etc should be correct.

dacarson commented:
> I could see if I can change the forecast fetching to the beginning of every hour rather than the end of every hour, that way as soon as it ticks over midnight, the forecast for Today, Tomorrow, etc should be correct.

Checking the code, I fetch the forecast as soon as we click over a new hour. (If the current hour has changed from the last time I fetched a forecast, then fetch a new forecast). This code is called at a 1 minute interval if you are using the Tempest weather station service. 
```
		if ((typeof this.lastForecastUpdate !== 'undefined')  && moment().hour() != this.lastForecastUpdate) {
			this.lastForecastUpdate = moment().hour();
... fetch forecast
```
I wonder if a new forecast hasn't been generated then. What I could do is if we are in a new hour, and if it is more than 2 minutes into that new hour, then fetch the forecast.
```
		if ((typeof this.lastForecastUpdate !== 'undefined')  && 
				(moment().hour() != this.lastForecastUpdate && moment().minute() > 2) {
			this.lastForecastUpdate = moment().hour();
... fetch forecast
```

wimleers commented:
> The `Observation` time in the screenshots are both before midnight: 22:59:36, and 23:59:36 so it thinks it is still on the same day. This will happen regardless of DST. On the next update, at 00:59:36, it should click over to the next day.

You misunderstood me: it was already wrong at 22:59:36: "Today" was still Friday! Same at 23:59:36. Only at 00:59:36 should it have said Saturday.


> DST is a tricky one. Sunrise and Sunset are computed using the same algorithm. Are they off by an hour too?

No, those are correct.

> I could see if I can change the forecast fetching to the beginning of every hour rather than the end of every hour, that way as soon as it ticks over midnight, the forecast for Today, Tomorrow, etc should be correct.

That shouldn't help: it was already saying "Today" was the next day between 22:00 and 23:00!

---
According to https://apidocs.tempestwx.com/reference/get_better-forecast the API response is of the format:
```
{
  "latitude": 29.00724,
  "longitude": -80.88067,
  "timezone": "America/New_York",
  "timezone_offset_minutes": -300,
  "location_name": "Home",
  "current_conditions": {
    "time": 1608668142,
    "conditions": "Partly Cloudy",
    "icon": "clear-day",
    "air_temperature": 67,
    "sea_level_pressure": 30.195,
    "station_pressure": 30.195,
    "pressure_trend": "falling",
    "relative_humidity": 55,
    "wind_avg": 3,
    "wind_direction": 15,
    "wind_direction_cardinal": "NNE",
    "wind_direction_icon": "wind-rose-sw",
    "wind_gust": 3,
    "solar_radiation": 830,
    "uv": 1,
    "brightness": 2582,
    "feels_like": 67,
    "dew_point": 67,
    "wet_bulb_temperature": 67,
    "wet_bulb_globe_temperature": 67,
    "delta_t": 10,
    "air_density": 0.08,
    "lighting_strike_count_last_1hr": 0,
    "lighting_strike_count_last_3hr": 2,
    "lighting_strike_last_distance": 7,
    "lighting_strike_last_distance_msg": "6 - 8 mi",
    "lighting_strike_last_epoch": 1608150959,
    "precip_accum_local_day": 0,
    "precip_accum_local_yesterday": 0,
    "precip_minutes_local_day": 0,
    "precip_minutes_local_yesterday": 0,
    "is_precip_local_day_rain_check": true,
    "is_precip_local_yesterday_rain_check": true,
    "includes_station_data": true
  },
  "forecast": {
    "daily": [
      {
        "day_start_local": 1608699600,
        "day_num": 5,
        "month_num": 7,
        "conditions": "Partly Cloudy",
        "icon": "clear-day",
        "sunrise": 1608639322,
        "sunset": 1608676362,
        "air_temp_high": 73,
        "air_temp_low": 54,
        "air_temp_high_color": "ff4c00",
        "air_temp_low_color": "afde06",
        "precip_probability": 10,
        "precip_icon": "chance-rain",
        "precip_type": "rain"
      }
    ],
    "hourly": [
      {
        "time": 1608735600,
        "conditions": "Partly Cloudy",
        "icon": "clear-day",
        "air_temperature": 67,
        "sea_level_pressure": 30.195,
        "relative_humidity": 55,
        "precip": 0.26,
        "precip_probability": 70,
        "precip_icon": "chance-rain",
        "wind_avg": 3,
        "wind_avg_color": "26fc59",
        "wind_direction": 15,
        "wind_direction_cardinal": "NNE",
        "wind_direction_icon": "wind-arrow-w",
        "wind_gust": 3,
        "wind_gust_color": "ebf100",
        "uv": 1,
        "feels_like": 67,
        "local_hour": 17,
        "local_day": 5
      }
    ]
  },
  "status": {
    "status_code": 0,
    "status_message": "Success"
  },
  "units": {
    "units_temp": "c",
    "units_wind": "mph",
    "units_precip": "in",
    "units_pressure": "mb",
    "units_distance": "string",
    "units_brightness": "lux",
    "units_solar_radiation": "W/m2",
    "units_other": "metric",
    "units_air_density": "kg/m3"
  },
  "source_id_conditions": 3
}
```

Looking at the PR, it does:
```
						weather.forecasts = this.parseForecasts(result["current_conditions"]["time"], result["forecast"]["daily"], result["timezone"]);
```
which in turn does:
```
			forecast.ForecastDay =  moment.unix(values[i].day_start_local).tz(timezone).format('dddd');
```
Either the `day_start_local` value in Weatherflow's API is wrong, or the `.tz()` functionality in `moment-timestamp` is wrong, because the forecast retrieved at 22:59 today on Sunday again indicates "Monday" as the "Today" forecast 🤷

OTOH, maybe … maybe it's just that the Weatherflow forecast simply does indeed forecast _the next day_ instead of _today_ as we near the end of the day? So perhaps the "Today" and "Tomorrow" labels are simply wrong? 🤔 
I'll add logging to see which exact timestamps are being provided by the API response, that'll clear this up! 👍

wimleers commented:
Dumped `results.forecast.daily`, got this:
```
3/11/2024, 11:03:03 PM] [homebridge-weather-plus] [
  {
    air_temp_high: 11,
    air_temp_low: 7,
    conditions: 'Rain Likely',
    day_num: 12,
    day_start_local: 1710198000,
    icon: 'rainy',
    month_num: 3,
    precip_icon: 'chance-rain',
    precip_probability: 90,
    precip_type: 'rain',
    sunrise: 1710223507,
    sunset: 1710265509
  },
  {
    air_temp_high: 13,
    air_temp_low: 10,
    conditions: 'Rain Possible',
    day_num: 13,
    day_start_local: 1710284400,
    icon: 'possibly-rainy-day',
    month_num: 3,
    precip_icon: 'chance-rain',
    precip_probability: 40,
    precip_type: 'rain',
    sunrise: 1710309773,
    sunset: 1710352010
  },
  {
    air_temp_high: 16,
    air_temp_low: 8,
    conditions: 'Partly Cloudy',
    day_num: 14,
    day_start_local: 1710370800,
    icon: 'partly-cloudy-day',
    month_num: 3,
    precip_icon: 'chance-rain',
    precip_probability: 10,
    precip_type: 'rain',
    sunrise: 1710396039,
    sunset: 1710438511
  },
  {
…
```

- Note the `day_num: 12` means "calendar day of the month", i.e. "March 12" right now
- Note the `1710198000` corresponds to 23:00 UTC on March 11, aka as 00:00 in `Europe/Brusssels`
- So … my hunch appears to have been correct. The very first entry in `result.forecast.daily` is the first available forecast, but is _not_ necessarily for today.

Therefore the solution is to:
1. verify that `day_start_local` corresponds to _today_'s start of the day (or tomorrow's, or … whichever "day + N" you enabled forecasts for
2. if no forecast is available for a certain day (e.g. today aka "day + 0"), then show the stale data if we have it or some kind of fault indicator otherwise

dacarson commented:
I get what you are saying. I **believe** that the same API is used to power the website for your weather station. 
Mine is https://tempestwx.com/station/39778/
If you change that last number to your station ID, can you check what it shows there? Because it seems very odd that they don't have the current day forecast when you are near the end of the day. What do they show on their site at the end of the day?

wimleers commented:
Will check tonight! 👍

dacarson commented:
Spending more time on the website I mentioned above, they don't actually show a **daily** forecast for `Today`, they only show an **hourly** forecast for `Today`. Therefore, I am not going to learn much about what happens at 2300-0000.

So, coming back to the solution, as you said:
> 1. verify that day_start_local corresponds to today's start of the day (or tomorrow's, or … whichever "day + N" you enabled forecasts for
> 2. if no forecast is available for a certain day (e.g. today aka "day + 0"), then show the stale data if we have it or some kind of fault indicator otherwise

I would need to slightly change this approach. The plugin loads the maximum number of forecast days, then only chooses to show the ones you have selected. Also, the plugin doesn't maintain stale forecast data. So what I could do is just populate the days that I get a forecast for, and any day I don't get a forecast for (ie the current day that is almost over), just leave it blank.

Looking at the network requests for the web page, I have confirmed that it does use pretty much the same [API](https://weatherflow.github.io/Tempest/api/swagger/#!/forecast/getBetterForecast) to fetch the forecast, and it gets practically the same response. It fetches it from https://swd.weatherflow.com/swd/rest/better_forecast?<parameters>

Patch coming

dacarson commented:
> Spending more time on the website I mentioned above, they don't actually show a **daily** forecast for `Today`, they only show an **hourly** forecast for `Today`. Therefore, I am not going to learn much about what happens at 2300-0000.

It actually does show the high/low temperature and the sunrise/sunset for the day. That said, I believe your solution still holds.

dacarson commented:
> My internet went down for a short period of time, got errors from other plugins but HomeBridge ended up restarting because of a change I made here [67c11a6](https://github.com/naofireblade/homebridge-weather-plus/commit/67c11a604a17f9ad9b4a9ca0e554cd384aa3e251). Patch coming

Patch is here (https://github.com/naofireblade/homebridge-weather-plus/pull/285/commits/d3a4f0b2cccaef044b5fe806611b476735a84bba)

dacarson commented:
Patch to address incorrect forecasts between 11pm and midnight (https://github.com/naofireblade/homebridge-weather-plus/pull/285/commits/df066d2e3d298eadf3658c3de1e512f4169ac459)
Check to make sure the forecast for today is actually today's forecast. If it is, then save a copy as I will need it later. When the forecast isn't today's, offset the forecast being stored into the forecasts array, and then insert the saved today's forecast. 

Tested by running over a couple of days and I saw that the Today's forecast had a different `Observation Time` than `In 1 Day`, also the `Day` values where correct.

@wimleers are you up for testing the fork again :-)

wimleers commented:
I've been logging the `results.forecast.daily` data. I've got enough data gathered now to confirm my earlier theory 👍 

- 00:00–22:00 on a certain calendar day:
```
[3/13/2024, 10:00:08 PM] [homebridge-weather-plus] [
  {
    air_temp_high: 13,
    air_temp_low: 10,
    conditions: 'Partly Cloudy',
    day_num: 13,
    day_start_local: 1710284400,
    icon: 'partly-cloudy-night',
    month_num: 3,
    precip_icon: 'chance-rain',
    precip_probability: 0,
    precip_type: 'rain',
…
```
- Starting at 23:00:
```
[3/13/2024, 11:00:09 PM] [homebridge-weather-plus] [
  {
    air_temp_high: 16,
    air_temp_low: 9,
    conditions: 'Rain Possible',
    day_num: 14,
    day_start_local: 1710370800,
    icon: 'possibly-rainy-day',
    month_num: 3,
    precip_icon: 'chance-rain',
    precip_probability: 20,
    precip_type: 'rain',
…
```

☝️ Once the last hour of the day has passed, there's no more forecast for _today_ available. Which kinda makes sense. And I bet that there's still an _hourly_ forecast for that last hour, but I did not check that.

dacarson commented:
> Once the last hour of the day has passed, there's no more forecast for today available. Which kinda makes sense. And I bet that there's still an hourly forecast for that last hour, but I did not check that.

You were right, thank you for debugging this. There is nothing in the API documentation about this behavior. On the website between 23:00-00:00 they drop the 'Today's' forecast and only show the forecast for the following day. 
This issue has now been fixed (https://github.com/naofireblade/homebridge-weather-plus/commit/df066d2e3d298eadf3658c3de1e512f4169ac459). 

I wonder if OpenWeatherMap has the same issue but no one has noticed it. I'll test later today.

wimleers commented:
Success! It's now 23:17, and your plugin is still showing Thursday for the `Today` forecast, with an `Observation time` of 22:00:02, but the `Tomorrow` forecast has been updated on 23:00:02 👍 

Let's ship this!

dacarson commented:
FWIW, OpenWeatherMap forecasts doesn't seem to have the same issue.


Yes, let's ship it. Thanks for all your help and support @wimleers

wimleers commented:
Still working flawlessly here 👍😊

dacarson commented:
Thanks @naofireblade! Just need to tag it as a new release :-) 

And thanks to @wimleers for all the testing and feedback.
I would like to close out as many open issues as possible so that there is a clean set of open issues remaining.

wimleers commented:
I'll try to chip in with that effort, @dacarson 😊 I know what a challenge that can be 🤓 For context, this is the open source project I work on for my `$DAYJOB`, and it's the same problem, but at quite a different scale: https://www.drupal.org/about/core/blog/bug-smash-initiative-3-year-update 😄

dacarson commented:
Thanks @wimleers 
All I can do is see if the issue has been fixed, or what the next steps are and add a note to the issue. In many cases the issue just needs to be closed as it behaves correctly, though I need to ask the reporter to do that.

naofireblade commented:
I just added you as a collaborator if you want to cleanup some old issues :) .

dacarson commented:
@naofireblade Thank you! I have started closing out old issues or issues that behave correctly.

dacarson commented:
@naofireblade I drafted a v3.3.4-Beta release tag. 
I was thinking of doing a Beta this time because then people can install it and try it out. Specifically the people who have filed issues that are fixed in this release can verify that they are fixed before I move it to an official release.

naofireblade commented:
Published to npm as 3.3.4-beta.1
Install with `npm install -g homebridge-weather-plus@beta`
